diff --git a/../brouter-master/brouter-core/src/main/java/btools/router/RoutingContext.java b/../brouter_akt/brouter-core/src/main/java/btools/router/RoutingContext.java
index 94fa329..65518b7 100644
--- a/../brouter-master/brouter-core/src/main/java/btools/router/RoutingContext.java
+++ b/../brouter_akt/brouter-core/src/main/java/btools/router/RoutingContext.java
@@ -122,6 +122,15 @@ public final class RoutingContext
   {
     BExpressionContext expctxGlobal = expctxWay; // just one of them...
 
+	if (keyValues != null) {
+		// add parameter to context
+		for (Map.Entry<String, String> e : keyValues.entrySet()) {
+			float f = Float.parseFloat(e.getValue());
+			expctxWay.setVariableValue(e.getKey(), f, false );
+			expctxNode.setVariableValue(e.getKey(), f, false );
+		}
+	}
+
     setModel( expctxGlobal._modelClass );
 
     downhillcostdiv = (int)expctxGlobal.getVariableValue( "downhillcost", 0.f );
